torchaudio/csrc/ray_tracing.cpp

```diff
@@ -0,0 +1,571 @@
+ * https://github.com/LCAV/pyroomacoustics
+ */
+#include <math.h>
+#include <torch/script.h>
+#include <torch/torch.h>
+
+namespace torchaudio {
+namespace rir {
+namespace {
+
+#define IS_HYBRID_SIM (false) // TODO: remove this once ISM method is supported
+#define ISM_ORDER (10) // TODO: remove this once ISM method is supported
+#define EPS ((scalar_t)(1e-5))
+#define VAL(t) (*((t).template data_ptr<scalar_t>()))
+
+/**
+ * Wall helper class. A wall records its own absorption, reflection and
+ * scattering coefficient, and exposes a few methods for geometrical operations
+ * (e.g. reflection of a ray)
+ */
+template <typename scalar_t>
+class Wall {
+ public:
+  Wall(
+      torch::Tensor _absorption,
+      torch::Tensor _scattering,
+      torch::Tensor _normal,
+      torch::Tensor _origin)
+      : absorption(_absorption),
+        reflection((scalar_t)1. - _absorption),
+        scattering(_scattering),
+        normal(_normal),
+        origin(_origin) {}
+
+  torch::Tensor get_absorption() {
+    return absorption;
+  }
+  torch::Tensor get_reflection() {
+    return reflection;
+  }
+  torch::Tensor get_scattering() {
+    return scattering;
+  }
+
+  /**
+   * Returns the side (-1, 1 or 0) on which a point lies w.r.t. the wall.
+   */
+  int side(torch::Tensor pos) {
+    auto dot = VAL((pos - origin).dot(normal));
+
+    if (dot > EPS) {
+      return 1;
+    } else if (dot < -EPS) {
+      return -1;
+    } else {
+      return 0;
+    }
+  }
+
+  /**
+   * Reflects a ray (dir) on the wall. Preserves norm of vector.
+   */
+  torch::Tensor reflect(torch::Tensor dir) {
+    return dir - normal * 2 * dir.dot(normal);
+  }
+
+  /**
+   * Returns the cosine angle of a ray (dir) with the normal of the wall
+   */
+  scalar_t cosine(torch::Tensor dir) {
+    return VAL(dir.dot(normal) / dir.norm());
+  }
+
+ private:
+  torch::Tensor absorption;
+  torch::Tensor reflection; // == 1 - absorption
+  torch::Tensor scattering;
+  torch::Tensor normal; // The normal to the wall: 2D or 3D vector
+  torch::Tensor
+      origin; // The origin of the wall: corresponds to an arbitrary corner.
+};
+
+/**
+ * RayTracer class helper for ray tracing. For attribute description, please see
+ * declarations below as well as Python wrapper.
+ */
+template <typename scalar_t>
+class RayTracer {
+ public:
+  RayTracer(
+      const torch::Tensor _room,
+      const torch::Tensor _source,
+      const torch::Tensor _mic_array,
+      int _num_rays,
+      const torch::Tensor _e_absorption,
+      const torch::Tensor _scattering,
+      scalar_t _mic_radius,
+      scalar_t _sound_speed,
+      scalar_t _energy_thres,
+      scalar_t _time_thres,
+      scalar_t _hist_bin_size)
+      : room(_room),
+        source(_source),
+        mic_array(_mic_array),
+        num_rays(_num_rays),
+        energy_0(2. / num_rays),
+        // e_absorption(_e_absorption),
+        // scattering(_scattering),
```

Comment:
why they are commented?

Reply:
They are leftovers from a previous implementation,  I moved them to the Wall class now. Thanks for flagging!